routing/opening_hours_ser_des.cpp

```diff
@@ -0,0 +1,358 @@
+            if (ShouldSkipRuleBecauseOldYear(range, m_currentYear))
+              oldRule = true;
+
+            item.SetYears({range});
+          });
+
+    apply(rules, rule.GetMonths(),
+          [&](osmoh::MonthdayRange const & range, osmoh::RuleSequence & item) {
+            if (ShouldSkipRuleBecauseOldYear(range, m_currentYear))
+              oldRule = true;
+
+            item.SetMonths({range});
+          });
+
+    if (oldRule)
+      continue;
+
+    apply(rules, rule.GetWeekdays().GetWeekdayRanges(),
+          [](osmoh::WeekdayRange const & range, osmoh::RuleSequence & item) {
+            osmoh::Weekdays weekdays;
+            weekdays.SetWeekdayRanges({range});
+            item.SetWeekdays(weekdays);
+          });
+
+    apply(rules, rule.GetWeekdays().GetHolidays(),
+          [](osmoh::Holiday const & holiday, osmoh::RuleSequence & item) {
+            auto weekdays = item.GetWeekdays();
+            weekdays.SetHolidays({holiday});
+            item.SetWeekdays(weekdays);
+          });
+
+    apply(rules, rule.GetTimes(),
+          [](osmoh::Timespan const & range, osmoh::RuleSequence & item) {
+            item.SetTimes({range});
+          });
+
+    apply(rules, std::vector<osmoh::RuleSequence::Modifier>{rule.GetModifier()},
+          [](osmoh::RuleSequence::Modifier modifier, osmoh::RuleSequence & item) {
+            item.SetModifier(modifier);
+          });
+
+    finalRules.insert(finalRules.end(), rules.begin(), rules.end());
+  }
+
+  std::vector<osmoh::RuleSequence> filteredRules;
+  for (auto rule : finalRules)
+  {
+    if (NotSupported(rule))
+      continue;
+
+    filteredRules.emplace_back(std::move(rule));
+  }
+
+  return filteredRules;
+}
+
+bool OpeningHoursSerDes::HaveSameHeaders(
+    std::vector<osmoh::RuleSequence> const & decomposedOhs) const
+{
+  CHECK(!decomposedOhs.empty(), ());
+  Header const header = CreateHeader(decomposedOhs.front());
+  for (auto const & oh : decomposedOhs)
+  {
+    if (header != CreateHeader(oh))
+      return false;
+  }
+
+  return true;
+}
+
+uint8_t OpeningHoursSerDes::GetBitsNumber(Header::Bits type) const
+{
+  switch (type)
+  {
+  case Header::Bits::Year: return 8;  // store value such that 2000 + value equals to real year
+  case Header::Bits::Month: return 4;
+  case Header::Bits::MonthDay: return 5;
+  case Header::Bits::WeekDay: return 3;
+  case Header::Bits::Hours: return 5;
+  case Header::Bits::Minutes: return 6;
+  case Header::Bits::Off: return 1;
+  case Header::Bits::Holiday: return 1;
+  case Header::Bits::Max: UNREACHABLE();
+  }
+  UNREACHABLE();
+}
+
+bool OpeningHoursSerDes::CheckYearRange(osmoh::MonthDay::TYear start,
+                                        osmoh::MonthDay::TYear end) const
+{
+  if (start < kYearBias || end < kYearBias)
+    return false;
+
+  // Should be filtered after |DecomposeOh| method.
+  CHECK(start >= m_currentYear || end >= m_currentYear, (start, end, m_currentYear));
```

Comment:
если подразумевать что эта сериализация будет использоваться не только для роутинга, то нужно описать какие правила мы выбрасываем:
- если в правиле упоминается год до 2000
- если правило действует после генерации данных

Reply:
> если правило действует после генерации данных

до генерации данных

Напишу в хедере класса

Reply:
спасибо!

Reply:
```cpp
/// \brief This class serialise opening hours from OSM format to some binary data.
/// Each opening hours has one, two, ... rules. Each rule has some data, like: Year, Month,
/// Weekdays etc. (look to |Header::Bits| for supported features). This class has two mods:
///     1) (default) Serialize opening hours only if all rules have same headers. In this case
///        we serialize in such format:
///        <number of rules><header><rule 1>...<rule N>
///     2) (Turn on with |SetSerializeEverything| method) For each rule serialize its' header.
///        Format:
///        <number of rules><header 1><rule 1><header 2><rule 2>...<header N><rule N>
///
/// First format can be used in routing for access:conditional because it has more strong filtering
/// of opening hours and the more trash will be dropped.
/// Second format can be used in opening hours for POI because it support more kinds of opening
/// hours and in case of invalid data we can show raw string and user will able to understand what
/// it's supposed to be.
```